// ng2-ckeditor/esm5/ckbutton.directive.js
import * as tslib_1 from "tslib";
import { Directive, EventEmitter, Output, Input } from '@angular/core';
/**
 * CKGroup component
 * Usage :
 *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
 *      <ckbutton [name]="'SaveButton'" [command]="'saveCommand'" (click)="save($event)"
 *                [icon]="'/save.png'" [toolbar]="'customGroup,1'" [label]="'Save'">
 *      </ckbutton>
 *   </ckeditor>
 */
import * as ɵngcc0 from '@angular/core';
var CKButtonDirective = /** @class */ (function () {
    function CKButtonDirective() {
        this.click = new EventEmitter();
    }
    CKButtonDirective.prototype.initialize = function (editor) {
        var _this = this;
        editor.instance.addCommand(this.command, {
            exec: function (evt) {
                _this.click.emit(evt);
            },
        });
        editor.instance.ui.addButton(this.name, {
            label: this.label,
            command: this.command,
            toolbar: this.toolbar,
            icon: this.icon,
        });
    };
    CKButtonDirective.prototype.ngOnInit = function () {
        if (!this.name)
            throw new Error('Attribute "name" is required on <ckbutton>');
        if (!this.command)
            throw new Error('Attribute "command" is required on <ckbutton>');
    };
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], CKButtonDirective.prototype, "click", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKButtonDirective.prototype, "label", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKButtonDirective.prototype, "command", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKButtonDirective.prototype, "toolbar", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKButtonDirective.prototype, "name", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKButtonDirective.prototype, "icon", void 0);
CKButtonDirective.ɵfac = function CKButtonDirective_Factory(t) { return new (t || CKButtonDirective)(); };
CKButtonDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CKButtonDirective, selectors: [["ckbutton"]], inputs: { label: "label", command: "command", toolbar: "toolbar", name: "name", icon: "icon" }, outputs: { click: "click" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CKButtonDirective, [{
        type: Directive,
        args: [{
                selector: 'ckbutton'
            }]
    }], function () { return []; }, { click: [{
            type: Output
        }], label: [{
            type: Input
        }], command: [{
            type: Input
        }], toolbar: [{
            type: Input
        }], name: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
    return CKButtonDirective;
}());
export { CKButtonDirective };
