// ng2-ckeditor/esm5/ckgroup.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, ContentChildren, QueryList } from '@angular/core';
import { CKButtonDirective } from './ckbutton.directive';
/**
 * CKGroup component
 * Usage :
 *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
 *      <ckgroup [name]="'exampleGroup2'" [previous]="'1'" [subgroupOf]="'exampleGroup1'">
 *          .
 *          .
 *      </ckgroup>
 *   </ckeditor>
 */
import * as ɵngcc0 from '@angular/core';
var CKGroupDirective = /** @class */ (function () {
    function CKGroupDirective() {
    }
    CKGroupDirective.prototype.ngAfterContentInit = function () {
        var _this = this;
        // Reconfigure each button's toolbar property within ckgroup to hold its parent's name
        this.toolbarButtons.forEach(function (button) { return (button.toolbar = _this.name); });
    };
    CKGroupDirective.prototype.initialize = function (editor) {
        editor.instance.ui.addToolbarGroup(this.name, this.previous, this.subgroupOf);
        // Initialize each button within ckgroup
        this.toolbarButtons.forEach(function (button) {
            button.initialize(editor);
        });
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKGroupDirective.prototype, "name", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CKGroupDirective.prototype, "previous", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CKGroupDirective.prototype, "subgroupOf", void 0);
    tslib_1.__decorate([
        ContentChildren(CKButtonDirective),
        tslib_1.__metadata("design:type", QueryList)
    ], CKGroupDirective.prototype, "toolbarButtons", void 0);
CKGroupDirective.ɵfac = function CKGroupDirective_Factory(t) { return new (t || CKGroupDirective)(); };
CKGroupDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CKGroupDirective, selectors: [["ckgroup"]], contentQueries: function CKGroupDirective_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, CKButtonDirective, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.toolbarButtons = _t);
    } }, inputs: { name: "name", previous: "previous", subgroupOf: "subgroupOf" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CKGroupDirective, [{
        type: Directive,
        args: [{
                selector: 'ckgroup'
            }]
    }], function () { return []; }, { name: [{
            type: Input
        }], previous: [{
            type: Input
        }], subgroupOf: [{
            type: Input
        }], toolbarButtons: [{
            type: ContentChildren,
            args: [CKButtonDirective]
        }] }); })();
    return CKGroupDirective;
}());
export { CKGroupDirective };
