// @busacca/ng-pick-datetime/esm2015/lib/date-time/adapter/moment-adapter/moment-date-time.module.js
/**
 * moment-date-time.module
 */
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { MomentDateTimeAdapter, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS } from './moment-date-time-adapter.class';
import { OWL_MOMENT_DATE_TIME_FORMATS } from './moment-date-time-format.class';
import { DateTimeAdapter, OWL_DATE_TIME_LOCALE } from '../date-time-adapter.class';
import { OWL_DATE_TIME_FORMATS } from '../date-time-format.class';
// import { DateTimeAdapter, OWL_DATE_TIME_FORMATS, OWL_DATE_TIME_LOCALE_PROVIDER } from 'ng-pick-datetime';
let MomentDateTimeModule = class MomentDateTimeModule {
};
MomentDateTimeModule = __decorate([
    NgModule({
        providers: [
            {
                provide: DateTimeAdapter,
                useClass: MomentDateTimeAdapter,
                deps: [OWL_DATE_TIME_LOCALE, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS]
            },
        ],
    })
], MomentDateTimeModule);
export { MomentDateTimeModule };
const ɵ0 = OWL_MOMENT_DATE_TIME_FORMATS;
let OwlMomentDateTimeModule = class OwlMomentDateTimeModule {
};
OwlMomentDateTimeModule = __decorate([
    NgModule({
        imports: [MomentDateTimeModule],
        providers: [{ provide: OWL_DATE_TIME_FORMATS, useValue: ɵ0 }],
    })
], OwlMomentDateTimeModule);
export { OwlMomentDateTimeModule };
export { ɵ0 };
